### ISSUE TITLE
chaincfg/chainhash: avoid allocations in calls to blake256

### ISSUE THREAD

**Issue by teknico** [decred/dcrd#1810]
With reference to [this Matrix message](https://matrix.to/#/!HEeJkbPRpAqgAwhXWO:decred.org/$15646751954108InVtG:decred.org?via=decred.org&via=matrix.org&via=zettaport.com).

Currently the dcrd `chainhash` package [calls](https://github.com/decred/dcrd/blob/master/chaincfg/chainhash/hashfuncs.go#L25) the `dchest/blake256` [package](https://github.com/dchest/blake256/blob/master/blake256.go#L161) via the `New` and `Sum` methods, that allocate memory on the heap.

The stdlib `sha256` package [uses](https://golang.org/src/crypto/sha256/sha256.go#273) a function, `Sum256`, that does not allocate memory. This API is also used by the btcsuite `chainhash` [package](https://github.com/btcsuite/btcd/blob/master/chaincfg/chainhash/hashfuncs.go#L23).

I will create the `Sum256` API in a fork of the `blake256` package. I will then change the dcrd `chainhash` package to use such an API, similarly to what btcsuite already does.

I do not have privileges to create the `decred/blake256` repo, so I will work in my own fork. Afterwards, someone with privileges should create the `decred/blake256` repo by forking my repo, and change my pull request so that the dcrd `chainhash` code uses `decred/blake256`.

**Comment by jrick** [comment]
It would be nice to get this upstreamed so others can benefit from it too.  Can you open an issue and/or pull request on dchest's repo for it?

**Comment by teknico** [comment]
As @davecgh mentioned on Matrix, I'd rather have our own fork in the Decred organization. That way we can make other changes (like adding a SIMD implementation) without depending on an external maintainer.

Others will still benefit from an actively maintained implementation.

And Decred seems one of the main users, most of the world moved to Blake2b/s a long time ago.

**Comment by davecgh** [comment]
I think it's a great idea to contribute back to upstream with the changes, but I would also prefer this live in the Decred org, specifically in the dcrd repository under a new `crypto/blake256` module.

My primary motivation for desiring that is that it is crucial to consensus, and we know from history that external projects typically do not have the same strict requirements that consensus requires (e.g. see the unintended Bitcoin fork as a result of OpenSSL).  This is why, in general, I prefer to house consensus critical code in the Decred org, especially if we're making modifications to it.  I am aware there are other cases of external dependencies in consensus critical code as well, such as `leveldb` and `siphash`, but that doesn't negate the desire in general to move towards avoiding external dependencies in consensus critical code as much as possible.

That said, I will concede that with the additional Go modules and the MVS it uses significantly improves the situation as compared to before since updating to new versions requires explicitly doing so (unless some transitive dependency ends up hoisting it forward, but this is easy to see from the module graph).  Nevertheless, I still think it's generally a better idea to be extremely paranoid, strict, and methodical about consensus dependencies.

**Comment by teknico** [comment]
Agreed, best having as few external dependencies as possible for consensus critical code.

I will both send the changes upstream and create a new `crypto/blake256` module in dcrd while changing the `chaincfg/chainhash` code.

**Comment by teknico** [comment]
Upstream merged my changes (except for one change that I'm going to keep in our copy, see the above discussion).

**Comment by jrick** [comment]
> My primary motivation for desiring that is that it is crucial to consensus, and we know from history that external projects typically do not have the same strict requirements that consensus requires (e.g. see the unintended Bitcoin fork as a result of OpenSSL). This is why, in general, I prefer to house consensus critical code in the Decred org, especially if we're making modifications to it. I am aware there are other cases of external dependencies in consensus critical code as well, such as leveldb and siphash, but that doesn't negate the desire in general to move towards avoiding external dependencies in consensus critical code as much as possible.

I agree with this sentiment in general, and especially in the case of leveldb, but cryptographic hashes are in their own category.  They must produce exactly the correct output and there is no room for differences in behavior without the entire value being invalid.  This is in contrast to something as complex as a database, where a bug could quite easily sneak in which results in a difficult-to-debug difference in the consensus validation.

> That said, I will concede that with the addition of Go modules and the MVS scheme they use significantly improves the situation as compared to before since updating to new versions requires explicitly doing so (unless some transitive dependency ends up hoisting it forward, but this is easy to see from the module graph). 

If they are our own dependencies which are pulling in a newer buggy version of some third party dependency, then we should fix that (and work with upstream to fix that too).  If we fork upstream but still use a dependency that pulls in the upstream, and upstream breaks, then that is still a possible consensus concern, so forking the repo is not enough -- it must be replaced everywhere as well.

**Comment by jrick** [comment]
On the topic of the optimizations themselves, while this is a nice change, we could also get most of the benefit by adding methods to wire.MsgTx to serialize the different prefix and witness serializations of a transaction.  Then for a task which requires much hashing, like building or validating merkle trees, a single hash.Hash could be reused over the entire calculation (or more if we save it) by using it as an io.Writer.

**Comment by dajohi** [comment]
put me down as an "against" forking blake256 until upstream rejects changes.

**Comment by davecgh** [comment]
> 
> 
> On the topic of the optimizations themselves, while this is a nice change, we could also get most of the benefit by adding methods to wire.MsgTx to serialize the different prefix and witness serializations of a transaction. Then for a task which requires much hashing, like building or validating merkle trees, a single hash.Hash could be reused over the entire calculation (or more if we save it) by using it as an io.Writer.

That doesn't work out in practice because calling `Sum` on the `hash.Hasher` results in 2 allocations even when reusing the hasher.  For reference, as the following benchmark of the code in #1809 shows, changing `blake256` with the suggested changes (as done in #1811) make a pretty massive difference:

```
benchmark                              old ns/op    new ns/op    delta
------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      23919        12450        -47.95%
BenchmarkCalcMerkleRootInPlace/1000    1099074      591386       -46.19%
BenchmarkCalcMerkleRootInPlace/2000    2210476      1178640      -46.68%
BenchmarkCalcMerkleRootInPlace/4000    4293553      2357785      -45.09%
BenchmarkCalcMerkleRootInPlace/8000    8808074      4711448      -46.51%
BenchmarkCalcMerkleRootInPlace/16000   17055449     9538869      -44.07%
BenchmarkCalcMerkleRootInPlace/32000   33706714     19291822     -42.77%

benchmark                              old allocs   new allocs   delta
-------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      64           0            -100.00%
BenchmarkCalcMerkleRootInPlace/1000    3004         0            -100.00%
BenchmarkCalcMerkleRootInPlace/2000    6004         0            -100.00%
BenchmarkCalcMerkleRootInPlace/4000    12004        0            -100.00%
BenchmarkCalcMerkleRootInPlace/8000    24004        0            -100.00%
BenchmarkCalcMerkleRootInPlace/16000   48004        0            -100.00%
BenchmarkCalcMerkleRootInPlace/32000   96004        0            -100.00%

benchmark                              old bytes    new bytes    delta
-------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      4432         0            -100.00%
BenchmarkCalcMerkleRootInPlace/1000    208272       0            -100.00%
BenchmarkCalcMerkleRootInPlace/2000    416272       0            -100.00%
BenchmarkCalcMerkleRootInPlace/4000    832272       0            -100.00%
BenchmarkCalcMerkleRootInPlace/8000    1664272      0            -100.00%
BenchmarkCalcMerkleRootInPlace/16000   3328272      0            -100.00%
BenchmarkCalcMerkleRootInPlace/32000   6656275      0            -100.00%
```

**Comment by davecgh** [comment]
I should also note that it also affects a lot of other code positively because it all goes through `chainhash` which means basically everything in the code base picks up the optimization.

**Comment by davecgh** [comment]
For further reference, here are the benchmarks of #1809 with avoiding `chainhash` and instead reusing a `hash.Hasher` as discussed above.  Not only is it worse performance as I called out above, but it ends up avoiding the existing infrastructure which is not ideal in my opinion either.

```
benchmark                              old ns/op    new ns/op    delta
------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      21570        17586        -18.47%
BenchmarkCalcMerkleRootInPlace/1000    1094095      826008       -24.50%
BenchmarkCalcMerkleRootInPlace/2000    2253597      1619452      -28.14%
BenchmarkCalcMerkleRootInPlace/4000    4511645      3419429      -24.21%
BenchmarkCalcMerkleRootInPlace/8000    9022715      6310216      -30.06%
BenchmarkCalcMerkleRootInPlace/16000   18849617     13292671     -29.48%
BenchmarkCalcMerkleRootInPlace/32000   35850716     26657934     -25.64%

benchmark                              old allocs   new allocs   delta
------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      64           44           -31.25%
BenchmarkCalcMerkleRootInPlace/1000    3004         2004         -33.29%
BenchmarkCalcMerkleRootInPlace/2000    6004         4004         -33.31%
BenchmarkCalcMerkleRootInPlace/4000    12004        8004         -33.32%
BenchmarkCalcMerkleRootInPlace/8000    24004        16004        -33.33%
BenchmarkCalcMerkleRootInPlace/16000   48004        32004        -33.33%
BenchmarkCalcMerkleRootInPlace/32000   96004        64004        -33.33%

benchmark                              old bytes    new bytes    delta
------------------------------------------------------------------------
BenchmarkCalcMerkleRootInPlace/20      4432         1552         -64.98%
BenchmarkCalcMerkleRootInPlace/1000    208272       64272        -69.14%
BenchmarkCalcMerkleRootInPlace/2000    416272       128272       -69.19%
BenchmarkCalcMerkleRootInPlace/4000    832272       256272       -69.21%
BenchmarkCalcMerkleRootInPlace/8000    664272       512272       -69.22%
BenchmarkCalcMerkleRootInPlace/16000   3328275      1024272      -69.23%
BenchmarkCalcMerkleRootInPlace/32000   6656275      2048273      -69.23%
```

**Comment by davecgh** [comment]
> I agree with this sentiment in general, and especially in the case of leveldb, but cryptographic hashes are in their own category. They must produce exactly the correct output and there is no room for differences in behavior without the entire value being invalid.

In theory, I agree.  Unfortunately, in practice, we've already seen that our `siphash` dependency (incidentally from the same org as the `blake256` dependency) was changed upstream and broke things on arm (even caused the code to panic) while it was still correct on other platforms.

**Comment by teknico** [comment]
Sorry for the negativity, but the discussion on my upstream PR doesn't exactly inspire confidence either.

**Comment by teknico** [comment]
I created another PR, #1814, that uses the new `blake256.Sum256` API from upstream, but without making a copy of the upstream code. Only one of #1811 and #1814 may go in, depending on the outcome of the discussion.

**Comment by davecgh** [comment]
Given the above discussion and reasoning, I'll be accepting #1811 once it's finalized.

**Comment by davecgh** [comment]
This was resolved via #1811 and #1818.
